### ISSUE TITLE
k-channel image dataset 

### ISSUE THREAD

**Issue by mtrth** [alexgkendall/caffe-segnet#8]
My dataset has 2 classes; with 1000 training images of (5,256,256) also corresponding ground truth data (1,256,256) which is a binary image either 0 or 1 to represent the 2 classes.
Can seg-net be used on k-channel dataset or is limited to RGB ?

**Comment by alexgkendall** [comment]
Yes, I've done something similar in the past with RGBD input images. You will need to package your data (5x256x256) into a format that Caffe can accept. For example, you can use Caffe's data layer to load your data from a LMDB database. If you can put your data in a matlab file (.mat) then [this script](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Scripts/convert_mat_to_lmdb.py) might help you to convert it to Caffe's LMDB format.

You then need to change the "DenseImageData" layer to two "Data" layers (one for your input, one for labels). Something like this:

layer {
  name: "data"
  type: "Data"
  top: "data"
  data_param {
    source: "name_of_your_lmdb_file"
    batch_size: 10
    backend: LMDB
  }
}

For the binary label data, you can follow the same setup that I describe in the [tutorial](http://mi.eng.cam.ac.uk/projects/segnet/tutorial.html).

Hope this helps!

**Comment by mtrth** [comment]
Hi I followed your suggestion

1.Modified convert_mat_to_lmdb.py added shuffle option and used tifffile package http://www.lfd.uci.edu/~gohlke/code/tifffile.py.html to read the images https://gist.github.com/mtrth/43b0f36dfd90ce4e6d10
1. Then in segnet_train.prototxt made changes

layer {
  name: "data"
  type: "Data"
  top: "data"
  data_param {
    source: "data_lmdb"    # Change this to the absolute path to your
    batch_size: 1  
   backend: LMDB
  }
}
layer {
  name: "data"
  type: "Data"
  top: "label"
  data_param {
    source: "labels"  
    batch_size: 1  
    backend: LMDB
  }
}
...
upsmaple4 changed 
    upsample_w: 32
    upsample_h: 32
...
upsample5 changed
    upsample_w: 16
    upsample_h: 16
...
in  conv1_2_D layer changed  num_output : 2 and also softmax layer
...
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "conv1_1_D"
  bottom: "label"
  top: "loss"
  loss_param: {
    weight_by_label_freqs: true
    class_weighting: 0.5
    class_weighting: 0.5
  }
}

 modified segnet_train.prototxt  https://gist.github.com/mtrth/c6e5851fae1e761fba11

but when I run ./SegNet/caffe-segnet/build/tools/caffe train -gpu 0 -solver /SegNet/Models/segnet_solver.prototxt  I get the error 

 I1120 15:39:08.722801  3215 solver.cpp:251] Learning Rate Policy: step
F1120 15:39:08.884181  3215 math_functions.cu:123] Check failed: status == CUBLAS_STATUS_SUCCESS (11 vs. 0)  CUBLAS_STATUS_MAPPING_ERROR

**Comment by alexgkendall** [comment]
Hi @mtrth ,

Almost - there are a few issues I have spotted at first glance:
1. In your script to convert images -> lmdb, Caffe needs the data with dimensions in this order: image,channel,width,height. So you will need to reshape your data to be in that order.
2. You shouldn't have to change conv1_2_D. You will need to modify conv1_1_D to output 2 channels (ie. num_output : 2).

Can you please give this a go and let me know if it works?

**Comment by mtrth** [comment]
hi @alexgkendall 
1) did a transpose so that dimensions are in order image,channel,width,height
final_data = np.transpose(final_data,(0,3,1,2)) 
final_label = np.transpose(final_label,(0,3,1,2))    

2)  I have changed  num_output in conv1_1_D ; made a mistake while mentioning here; but the modified segnet_train.prototxt https://gist.github.com/mtrth/c6e5851fae1e761fba11 

Its training now thanks for the help. 
But the accuracy seems to oscillate and per_class_accuracy is sometimes -nan ; should I adjust the base_lr and momentum parameter ?

Also could you explain the class_weighting and weight_by_label_freqs parameter in the prototxt file and how it will influence my training. 

I1121 12:17:50.055621  9402 solver.cpp:214] Iteration 0, loss = 0.200133
I1121 12:17:50.055665  9402 solver.cpp:229]     Train net output #0: accuracy = 0.836365
I1121 12:17:50.055676  9402 solver.cpp:229]     Train net output #1: loss = 0.200133 (\* 1 = 0.200133 loss)
I1121 12:17:50.055685  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.0714286
I1121 12:17:50.055691  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.836855
I1121 12:17:50.055717  9402 solver.cpp:486] Iteration 0, lr = 0.001
I1121 12:17:56.099216  9402 solver.cpp:214] Iteration 20, loss = 0.683113
I1121 12:17:56.099261  9402 solver.cpp:229]     Train net output #0: accuracy = 0.194595
I1121 12:17:56.099272  9402 solver.cpp:229]     Train net output #1: loss = 0.683113 (\* 1 = 0.683113 loss)
I1121 12:17:56.099278  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.194595
I1121 12:17:56.099284  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = -nan
I1121 12:17:56.099290  9402 solver.cpp:486] Iteration 20, lr = 0.001
I1121 12:18:02.153717  9402 solver.cpp:214] Iteration 40, loss = 0.356544
I1121 12:18:02.153761  9402 solver.cpp:229]     Train net output #0: accuracy = 0.585464
I1121 12:18:02.153771  9402 solver.cpp:229]     Train net output #1: loss = 0.356544 (\* 1 = 0.356544 loss)
I1121 12:18:02.153779  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.586568
I1121 12:18:02.153784  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.38674
I1121 12:18:02.153790  9402 solver.cpp:486] Iteration 40, lr = 0.001
I1121 12:18:08.202131  9402 solver.cpp:214] Iteration 60, loss = 0.556992
I1121 12:18:08.202155  9402 solver.cpp:229]     Train net output #0: accuracy = 0.329056
I1121 12:18:08.202164  9402 solver.cpp:229]     Train net output #1: loss = 0.556992 (\* 1 = 0.556992 loss)
I1121 12:18:08.202172  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.204494
I1121 12:18:08.202177  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.791104
I1121 12:18:08.202183  9402 solver.cpp:486] Iteration 60, lr = 0.001
I1121 12:18:14.245791  9402 solver.cpp:214] Iteration 80, loss = 0.654159
I1121 12:18:14.245817  9402 solver.cpp:229]     Train net output #0: accuracy = 0.212158
I1121 12:18:14.245827  9402 solver.cpp:229]     Train net output #1: loss = 0.654159 (\* 1 = 0.654159 loss)
I1121 12:18:14.245833  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = -nan
I1121 12:18:14.245839  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.212158
I1121 12:18:14.245846  9402 solver.cpp:486] Iteration 80, lr = 0.001
I1121 12:18:20.289883  9402 solver.cpp:214] Iteration 100, loss = 0.702287
I1121 12:18:20.289962  9402 solver.cpp:229]     Train net output #0: accuracy = 0.160263
I1121 12:18:20.289971  9402 solver.cpp:229]     Train net output #1: loss = 0.702287 (\* 1 = 0.702287 loss)
I1121 12:18:20.289978  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.160263
I1121 12:18:20.289983  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = -nan
I1121 12:18:20.289990  9402 solver.cpp:486] Iteration 100, lr = 0.001
I1121 12:18:26.332846  9402 solver.cpp:214] Iteration 120, loss = 0.251738
I1121 12:18:26.332869  9402 solver.cpp:229]     Train net output #0: accuracy = 0.766464
I1121 12:18:26.332878  9402 solver.cpp:229]     Train net output #1: loss = 0.251738 (\* 1 = 0.251738 loss)
I1121 12:18:26.332885  9402 solver.cpp:229]     Train net output #2: per_class_accuracy = -nan
I1121 12:18:26.332890  9402 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.766464

**Comment by alexgkendall** [comment]
Fantastic!

If per_class_accuracy is nan, it means that that class wasn't present in the training mini-batch. Accuracy is computed as %true positives for that class, hence when that class isn't present, it is dividing by zero. You can ignore these lines.

If you assign a weight by label freq, it multiplies the loss for that pixel by the scaling factor for the given ground truth class. See our [SegNet arXiv paper](http://arxiv.org/abs/1511.00561) for the formula we use to weight the classes during training. If you want the same learning rate for each class (ie. what you do here) then you do not need to specify a class weighting. SegNet will use a weighting value of 1 for each class as default. So you can remove these three lines:
    weight_by_label_freqs: true
    class_weighting: 0.5
    class_weighting: 0.5

Your problem is now a learning problem, your model is not converging. What batch size are you using? What did you observe after leaving it learning for a while (eg. approx 1000 iterations)? It will be noisy over a very small window such as the 100 iterations you have shown. Typically on the datasets I work with, SegNet converges after 100 epochs. I would recommend trying the following:
- lower the learning rate until it starts to converge
- try a smaller model, such as SegNet basic?

**Comment by mtrth** [comment]
hi @alexgkendall 
the batch size is 4 , and I also removed the loss_param in SoftmaxWithLoss; I experimented with different learning rates for both SegNet basic and SegNet
SegNet ( learning rate: 0.0001, 0.00001)
SegNet basic ( learning rate: 0.001, 0.0001, 0.00001)

I found SegNet basic with learning rate 0.00001 to be best the complete training log is https://gist.github.com/mtrth/4153454a4a30ff5af7be ; would this be the optimal parameter to proceed with?

Also following your tutorial in the Testing SegNet portion when I run the script compute_bn_statistics.py it throughs a assertion error  as I am passing lmdb at the data layer,
 assert net_message.layer[0].type == "DenseImageData"

**Comment by alexgkendall** [comment]
Cool. If you have a validation set then the optimal model would be the one which performs best on this data. When you say best, do you mean it has the lowest training loss? This should roughly correlate with best performing model. If you do observe it overfitting then you can try introducing some dropout layers.

That test script is set up to deal with the "DenseImageData" layer with image files. Since you are using "Data" layer with LMDB files then you will need to modify the script.

**Comment by mtrth** [comment]
Yes I had tested it on a validation set and it gave the lowest training loss. I also modified test script to deal with LMDB data; the segmentation results on my sample test set are good with around 85% accuracy. Thank you for the all help.

**Comment by alexgkendall** [comment]
Awesome that's very exciting - what sort of data are you working with? It would be cool to see some results!

**Comment by mtrth** [comment]
Hi Alex, I have sent a follow up email about the results to your id agk34@cam.ac.uk

**Comment by cslxiao** [comment]
hi, @mtrth, I am conducting the same experiment which uses LMDB data. 
How did you modify the test script to deal with LMDB data?  I am not familiar with python. Would you share it with me? I appreciate your help.

**Comment by Soham0** [comment]
hi @alexgkendall What is relation between number of iterations and epochs in the case of SegNet?
Also while setting the lr for certain layers to 0, is it required to be done for encoder as well as decoder layers?

**Comment by leihamilton** [comment]
Hi, @alexgkendall @mtrth , I would like to try RGBD data on SegNet, I followed your steps and created a lmdb database for RGBD, and a lmdb database for label. In the original SegNet, the data is first pasted into the NORM layer, but in @mtrth posted code, it skip the NORM. My questions is that do I still pass the RGBD to the NORM layer? Thanks.

**Comment by zhuanjiao2222** [comment]
Hi, @leihamilton , I'm now trying to create a lmdb database for RGBD, could you tell how to convert images to a lmdb database?

**Comment by leihamilton** [comment]
@zhuanjiao2222 , here is an example, https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Scripts/convert_mat_to_lmdb.py. You need to mordify the code a little if the input is image.

**Comment by zhuanjiao2222** [comment]
Hi, @leihamilton, thank you for your help, I have converted images to lmdb format successfully, but now I have some new questions.
1. Caffe uses BGR format, when I use Images as input to train the net, whether the image is converted to BGR format by caffe automatically. The original image is RGB format.
2. If I use lmdb database as input to train the net, whether I should convert the image to lmdb format in BGR or BGR-D order?
